### ISSUE TITLE
Shader error on Yosemite (AMD Radeon R9 M370X)

### ISSUE THREAD

**Issue by liolaarc** [afainer/cleven#1]
Hi there,

See https://news.ycombinator.com/item?id=10653568

I ran into a shader error when loading the engine:

```
  (ql:quickload "sdl2")
  (ql:quickload "cleven")
  (load "sandbox.lisp")
  (in-package #:sandbox)
  (run-sandbox)

  debugger invoked on a SIMPLE-ERROR in thread
  #<THREAD "Render thread" RUNNING {1006D70EF3}>:
    Shader compilation failed.  The info log is:
  ERROR: 0:24: '' :  version '330' is not supported
```

I tested it on another computer and got the same error.

How would you debug this or look into it further?  Is there any useful info I should send to you?  (Debug logs, system specs, etc.)

Have a good week!

**Comment by PuercoPop** [comment]
I've have a similar issue in other projects, afaiu the problem is that the gl context does not default to 330 in the card even though the card supports it. Try adding

```
  (sdl2:gl-set-attr :context-major-version 3)
  (sdl2:gl-set-attr :context-minor-version 3)
```

Before making the gl context, [here](https://github.com/afainer/cleven/blob/cf7a5a00684e268c659418832efdb7bd8eda87cd/game.lisp#L99)

Afterwards I get a different error

```
OpenGL signalled (1280 . INVALID-ENUM) from ENABLE.
```

**Comment by liolaarc** [comment]
I see it too.  Thanks!  Your code almost worked, but here's the full code I ended up needing:

``` lisp

(defun run-game (sdl-init-flags
                 &key (win-title "Cleven game")
                   (win-x :centered) (win-y :centered)
                   (win-w 800) (win-h 600)
                   (win-flags (list :shown :opengl)))
  "Run game with SDL-INIT-FLAGS and optional window parameters."
  (with-game-init sdl-init-flags
    ;; TODO It would be nice to have ability to create multiple
    ;; windows
    (sdl2:with-window (win
                       :title win-title
                       :x win-x :y win-y
                       :w win-w :h win-h
                       :flags win-flags)
      (unwind-protect
           (progn
             (sdl2:gl-set-attr :context-major-version 3)
             (sdl2:gl-set-attr :context-minor-version 3)
             (sdl2:gl-set-attr :context-profile-mask #x1)
             (lock #'run-hook-with-args *game-init-functions-lock* win)
             (event-loop))
        (quit-all-threads)
        ;; It is save to run quit hooks now without locking
        (run-hooks '*game-quit-hook*)))))
```

The `(sdl2:gl-set-attr :context-profile-mask #x1)` was necessary due to http://stackoverflow.com/a/13095742

Try running `sbcl --disable-debugger` at startup and you'll get a couple stack traces:

```
* (in-package #:sandbox)

#<PACKAGE "CLEVEN-SANDBOX">
* (run-sandbox)
Unhandled CL-OPENGL-BINDINGS:OPENGL-ERROR in thread #<SB-THREAD:THREAD
                                                      "Render thread" RUNNING
                                                       {10080C8F53}>:
  OpenGL signalled (1280 . INVALID-ENUM) from ENABLE.

Backtrace for: #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}>
0: ((LAMBDA NIL :IN SB-DEBUG::FUNCALL-WITH-DEBUG-IO-SYNTAX))
1: (SB-IMPL::CALL-WITH-SANE-IO-SYNTAX #<CLOSURE (LAMBDA NIL :IN SB-DEBUG::FUNCALL-WITH-DEBUG-IO-SYNTAX) {10080D84DB}>)
2: (SB-IMPL::%WITH-STANDARD-IO-SYNTAX #<CLOSURE (LAMBDA NIL :IN SB-DEBUG::FUNCALL-WITH-DEBUG-IO-SYNTAX) {10080D84AB}>)
3: (PRINT-BACKTRACE :STREAM #<SYNONYM-STREAM :SYMBOL SB-SYS:*STDERR* {10001527E3}> :START 0 :FROM :INTERRUPTED-FRAME :COUNT NIL :PRINT-THREAD T :PRINT-FRAME-SOURCE NIL :METHOD-FRAME-STYLE NIL)
4: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<CL-OPENGL-BINDINGS:OPENGL-ERROR {10080D3983}> #<unavailable argument>)
5: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<CL-OPENGL-BINDINGS:OPENGL-ERROR {10080D3983}>)
6: (INVOKE-DEBUGGER #<CL-OPENGL-BINDINGS:OPENGL-ERROR {10080D3983}>)
7: (ERROR #<CL-OPENGL-BINDINGS:OPENGL-ERROR {10080D3983}>)
8: (CL-OPENGL-BINDINGS:CHECK-ERROR #<unavailable argument>)
9: ((LAMBDA NIL :IN CLEVEN:RENDER-LOOP))
10: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
11: ((FLET #:WITHOUT-INTERRUPTS-BODY-1131 :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
12: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
13: ((FLET #:WITHOUT-INTERRUPTS-BODY-578 :IN SB-THREAD::CALL-WITH-MUTEX))
14: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {CE2DCFB}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}>> NIL T NIL)
15: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}> NIL #<CLOSURE (LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS) {10080C8EFB}> (#<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}> #<SB-THREAD:THREAD "SDL2 Main Thread" RUNNING {10080C7983}> #<SB-THREAD:THREAD "main thread" waiting on: #<WAITQUEUE Anonymous condition variable {10080C7BB3}> {1002D86A73}>) NIL NIL NIL NIL)
16: ("foreign function: call_into_lisp")
17: ("foreign function: new_thread_trampoline")
18: ("foreign function: _pthread_body")
19: ("foreign function: _pthread_body")
20: ("foreign function: thread_start")

unhandled condition in --disable-debugger mode, quitting
Unhandled TYPE-ERROR in thread #<SB-THREAD:THREAD "Render thread" RUNNING
                                  {10080C8F53}>:
  The value NIL is not of type (UNSIGNED-BYTE 32).

Backtrace for: #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}>
0: (CL-OPENGL-BINDINGS:GET-PROGRAM-IV NIL :ATTACHED-SHADERS #.(SB-SYS:INT-SAP #X00353DD0))
1: (CL-OPENGL::GET-PROGRAM-AUX NIL :ATTACHED-SHADERS :INT 1)
2: (CL-OPENGL:GET-ATTACHED-SHADERS NIL)
3: (CLEVEN::FREE-SHADERS NIL)
4: (CLEVEN::RENDER-FREE-SHADERS)
5: (CLEVEN::RUN-HOOKS CLEVEN:*RENDER-QUIT-HOOK*)
6: ((FLET #:WITHOUT-INTERRUPTS-BODY-578 :IN SB-THREAD::CALL-WITH-MUTEX))
7: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN CLEVEN:LOCK) {CE2D0DB}> #<SB-THREAD:MUTEX "Anonymous lock" owner: #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}>> NIL T NIL)
8: (CLEVEN:LOCK #<FUNCTION CLEVEN::RUN-HOOKS> (#<SB-THREAD:MUTEX "Anonymous lock" owner: #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}>> . CLEVEN:*RENDER-QUIT-HOOK*))
9: ((FLET #:CLEANUP-FUN-22 :IN CLEVEN:RENDER-LOOP)) [cleanup]
10: ((LAMBDA NIL :IN CLEVEN:RENDER-LOOP))
11: ((LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS))
12: ((FLET #:WITHOUT-INTERRUPTS-BODY-1131 :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
13: ((FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE))
14: ((FLET #:WITHOUT-INTERRUPTS-BODY-578 :IN SB-THREAD::CALL-WITH-MUTEX))
15: (SB-THREAD::CALL-WITH-MUTEX #<CLOSURE (FLET SB-THREAD::WITH-MUTEX-THUNK :IN SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE) {CE2DCFB}> #<SB-THREAD:MUTEX "thread result lock" owner: #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}>> NIL T NIL)
16: (SB-THREAD::INITIAL-THREAD-FUNCTION-TRAMPOLINE #<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}> NIL #<CLOSURE (LAMBDA NIL :IN BORDEAUX-THREADS::BINDING-DEFAULT-SPECIALS) {10080C8EFB}> (#<SB-THREAD:THREAD "Render thread" RUNNING {10080C8F53}> #<SB-THREAD:THREAD "SDL2 Main Thread" RUNNING {10080C7983}> #<SB-THREAD:THREAD "main thread" waiting on: #<WAITQUEUE Anonymous condition variable {10080C7BB3}> {1002D86A73}>) NIL NIL NIL NIL)
17: ("foreign function: call_into_lisp")
18: ("foreign function: new_thread_trampoline")
19: ("foreign function: _pthread_body")
20: ("foreign function: _pthread_body")
21: ("foreign function: thread_start")

unhandled condition in --disable-debugger mode, quitting
```

Sadly, that's as far as I can go with this for now.  Hopefully the author will be able to take it from here, or suggest next steps.  I used to be able to debug GL errors effectively, but... well... Time passed. :)

**Comment by afainer** [comment]
This is strange! As far аs I can see [here](http://www.amd.com/en-us/products/graphics/notebook/r9-m200#) your video card supports OpenGL 4.3.  Maybe your driver is not backward compatible with OpenGL 3.3.  Anyway, try a quick fix.  Just change the version in vertex and fragment shaders to 430.

-#version 330
+#version 430

BTW, I'm running the project on Linux with Nvidia GeForce GT 720M.

**Comment by liolaarc** [comment]
The issue at this point is that on OS X, some of your GL calls generate an
invalid enum error.  These can be tricky to track down, because pretty much
any GL call can generate them.  You may want to try to wrap all your GL
statements in some kind of glGetError() call; at that point, I'll be able
to run the engine and tell you which call is failing, which will lead
directly to finding out why.

Unfortunately, a sequence of operations that are valid on Linux can cause
GL errors on other platforms, due to driver differences.  But in this case,
I suspect it's an NVidia vs AMD issue.  Different brands of video cards
will generate widely different behaviors, so setting up the engine to call
glGetError() after every GL call is pretty important.  (Unfortunately, the
nature of GL means this isn't too easy.  Maybe SDL2's CL wrappers have some
kind of option to handle this for you, though you'll probably want to set
up the equivalent yourself for generality.)

Best of luck!

**Comment by afainer** [comment]
Thanks for your comments!

My experience with OpenGL is quite limited, so your comments are very
useful.  But I think the issue might be the `render-loop`.
Could you please add debug optimization before its definition.

``` Lisp
(declaim (optimize (debug 3)))
(defun render-loop (win)
...
```

Then load the Cleven system

``` Lisp
(asdf:load-system :cleven)
```

and run the Cleven sandbox.  You should get more informative backtrace.

Please, try to play with code before `unwind-protect`.  May be
something wrong with it

``` Lisp
(gl:disable :dither)
(gl:enable :alpha-test)
(gl:alpha-func :greater 0)
(gl:enable :blend)
(gl:blend-equation :func-add)
(gl:blend-func :one :one-minus-src-alpha)
```

Thank you and kind regards!

**Comment by seanjensengrey** [comment]
On 

```
Mac OS X 10.10.4
AMD Radeon HD 6750M 1024 MB
```

these changes do not fix the original problem and introduce other memory faults. 

```
CORRUPTION WARNING in SBCL pid 73775(tid 2953326592):
Memory fault at 0x0 (pc=0x7fff8e027534, sp=0x81ff4d0)
The integrity of this image is possibly compromised.
Continuing with fingers crossed.

debugger invoked on a SB-SYS:MEMORY-FAULT-ERROR in thread
#<THREAD "SDL2 Main Thread" RUNNING {100A1B55D3}>:
  Unhandled memory fault at #x0.
```

I think laarc 's advice in wrapping OpenGL to call glGetError() as the best path forward.
